Cannot accelerate expressions containing `Constant`s with Pyccel

When discretizing a `BilinearForm`, `LinearForm` or `Functional` with a Pyccel backend, Psydac cannot generate the correct type annotation for a function argument of type `Constant`. This results in the error

```python
    def build_pyccel_types_decorator(args, order=None):
        """
        builds a types decorator from a list of arguments (of FunctionDef)
        """
        types = []
        for a in args:
            if isinstance(a, Variable):
                dtype = a.dtype.name.lower()
    
            elif isinstance(a, IndexedVariable):
                dtype = a.dtype.name.lower()
    
            else:
>               raise TypeError('unexpected type for {}'.format(a))
E               TypeError: unexpected type for c

../ast/utilities.py:831: TypeError
```

--- a/psydac/api/ast/utilities.py
+++ b/psydac/api/ast/utilities.py
@@ -21,6 +21,7 @@ from sympde.topology.derivatives import get_atom_logical_derivatives
 from sympde.topology.derivatives import get_index_logical_derivatives
 from sympde.topology             import LogicalExpr
 from sympde.topology             import SymbolicExpr
+from sympde.core                 import Constant
 
 from pyccel.ast.core import Variable, IndexedVariable
 from pyccel.ast.core import For
@@ -822,19 +823,31 @@ def build_pyccel_types_decorator(args, order=None):
     types = []
     for a in args:
         if isinstance(a, Variable):
+            rank  = a.rank
             dtype = a.dtype.name.lower()
 
         elif isinstance(a, IndexedVariable):
+            rank  = a.rank
             dtype = a.dtype.name.lower()
+
+        elif isinstance(a, Constant):
+            rank = 0
+            if a.is_integer:
+                dtype = 'int'
+            elif a.is_complex:
+                dtype = 'complex'
+            elif a.is_real:
+                dtype = 'float'
+            else:
+                dtype = 'float' # default value
 
         else:
             raise TypeError('unexpected type for {}'.format(a))
 
-        if a.rank > 0:
-            shape = [':' for i in range(0, a.rank)]
-            shape = ','.join(i for i in shape)
+        if rank > 0:
+            shape = ','.join(':' * rank)
             dtype = '{dtype}[{shape}]'.format(dtype=dtype, shape=shape)
-            if order and a.rank > 1:
+            if order and rank > 1:
                 dtype = "{dtype}(order={ordering})".format(dtype=dtype, ordering=order)
 
         dtype = String(dtype)

--- a/psydac/api/tests/test_assembly.py
+++ b/psydac/api/tests/test_assembly.py
@@ -1,3 +1,5 @@
+import pytest
+
 from sympde.topology import Line, Square
 from sympde.topology import ScalarFunctionSpace
 from sympde.topology import element_of
@@ -8,9 +10,19 @@ from sympde.expr     import integral
 
 from psydac.api.discretization import discretize
 from psydac.fem.basic          import FemField
+from psydac.api.settings       import PSYDAC_BACKENDS
 
 #==============================================================================
-def test_field_and_constant():
+@pytest.fixture(params=[None, 'numba', 'pyccel-gcc'])
+def backend(request):
+    return request.param
+
+#==============================================================================
+def test_field_and_constant(backend):
+
+    # If 'backend' is specified, accelerate Python code by passing **kwargs
+    # to discretization of bilinear forms, linear forms and functionals.
+    kwargs = {'backend': PSYDAC_BACKENDS[backend]} if backend else {}
 
     domain = Square()
     V = ScalarFunctionSpace('V', domain)
@@ -27,8 +39,8 @@ def test_field_and_constant():
     degree = (3, 3)
     domain_h = discretize(domain, ncells=ncells)
     Vh = discretize(V, domain_h, degree=degree)
-    ah = discretize(a, domain_h, [Vh, Vh])
-    lh = discretize(l, domain_h, Vh)
+    ah = discretize(a, domain_h, [Vh, Vh], **kwargs)
+    lh = discretize(l, domain_h,      Vh , **kwargs)
 
     fh = FemField(Vh)
     fh.coeffs[:] = 1
@@ -47,7 +59,11 @@ def test_field_and_constant():
     print("PASSED")
 
 #==============================================================================
-def test_multiple_fields():
+def test_multiple_fields(backend):
+
+    # If 'backend' is specified, accelerate Python code by passing **kwargs
+    # to discretization of bilinear forms, linear forms and functionals.
+    kwargs = {'backend': PSYDAC_BACKENDS[backend]} if backend else {}
 
     domain = Line()
     V = ScalarFunctionSpace('V', domain)
@@ -65,8 +81,8 @@ def test_multiple_fields():
     degree = (3,)
     domain_h = discretize(domain, ncells=ncells)
     Vh = discretize(V, domain_h, degree=degree)
-    ah = discretize(a, domain_h, [Vh, Vh])
-    lh = discretize(l, domain_h, Vh)
+    ah = discretize(a, domain_h, [Vh, Vh], **kwargs)
+    lh = discretize(l, domain_h,      Vh , **kwargs)
 
     fh = FemField(Vh)
     fh.coeffs[:] = 1

